{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3c3113ec-6279-46b0-b266-ae28c124c0f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pandas as pd\n",
    "import random\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b5463caa-9812-4853-a94b-fe7757988668",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>frame</th>\n",
       "      <th>strand</th>\n",
       "      <th>protein</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>107</td>\n",
       "      <td>500</td>\n",
       "      <td>393</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>MVFSIIATRWLVILAGAERTVATCPCLALLSRISATRRKRSAFATD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>336</td>\n",
       "      <td>2799</td>\n",
       "      <td>2463</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2800</td>\n",
       "      <td>3733</td>\n",
       "      <td>933</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>MVKVYAPASSANMSVGFDVLGAAVTPVDGALLGDVVTVEAAETFSL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3511</td>\n",
       "      <td>4162</td>\n",
       "      <td>651</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>MSHCRSGITGGRGQNGHWLITRNMGQHLCHKATAEIFKCQRWPVEQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3733</td>\n",
       "      <td>5020</td>\n",
       "      <td>1287</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>MKLYNLKDHNEQVSFAQAVTQGLGKNQGLFFPHDLPEFSLTEIDEM...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start   end  length  frame strand  \\\n",
       "0    107   500     393      2      -   \n",
       "1    336  2799    2463      0      +   \n",
       "2   2800  3733     933      1      +   \n",
       "3   3511  4162     651      0      -   \n",
       "4   3733  5020    1287      1      +   \n",
       "\n",
       "                                             protein  \n",
       "0  MVFSIIATRWLVILAGAERTVATCPCLALLSRISATRRKRSAFATD...  \n",
       "1  MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...  \n",
       "2  MVKVYAPASSANMSVGFDVLGAAVTPVDGALLGDVVTVEAAETFSL...  \n",
       "3  MSHCRSGITGGRGQNGHWLITRNMGQHLCHKATAEIFKCQRWPVEQ...  \n",
       "4  MKLYNLKDHNEQVSFAQAVTQGLGKNQGLFFPHDLPEFSLTEIDEM...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orfs = pd.read_csv(\"../results/ecoli_orfs_v2.csv\")\n",
    "orfs.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7ee93a71-524d-4f53-8f9f-c1d1e50c4028",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(42)\n",
    "sampled_orfs = orfs.sample(50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "29480edc-fc67-4a03-b6e1-43557b9aae5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "orfs = orfs[orfs[\"length\"] >= 300]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6593dc50-ee6c-4bd2-857c-f07ad9d85c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = next(SeqIO.parse(\"../data/genomes/ecoli.fasta\", \"fasta\"))\n",
    "genome_seq = genome.seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3d2264e7-69d2-438e-81a5-568d20bf25ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_records = []\n",
    "\n",
    "for i, row in sampled_orfs.iterrows():\n",
    "    seq = genome_seq[row[\"start\"]:row[\"end\"]]\n",
    "    record = SeqIO.SeqRecord(\n",
    "        seq,\n",
    "        id=f\"gene_{i}\",\n",
    "        description=f\"start={row['start']},end={row['end']}\"\n",
    "    )\n",
    "    gene_records.append(record)\n",
    "\n",
    "SeqIO.write(gene_records, \"../data/genes/ecoli_50_genes.fasta\", \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "807b336c-eaef-4230-863f-7be558383078",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = [\n",
    "    \"qseqid\",\"sseqid\",\"pident\",\"length\",\"mismatch\",\"gapopen\",\n",
    "    \"qstart\",\"qend\",\"sstart\",\"send\",\"evalue\",\"bitscore\"\n",
    "]\n",
    "\n",
    "shi = pd.read_csv(\"../results/ecoli_vs_shigella.tsv\", sep=\"\\t\", names=cols)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8136268e-0f9e-41d6-ae3d-91edd9a99480",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_hits(df):\n",
    "    return df[\n",
    "        (df[\"evalue\"] < 1e-10) &\n",
    "        (df[\"pident\"] >= 70)\n",
    "    ]\n",
    "\n",
    "shi_f = filter_hits(shi)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "db49f850-2770-4134-8ed2-670994978fc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_shi = set(shi_f[\"qseqid\"])\n",
    "genes_all = set(sampled_orfs.index.map(lambda x: f\"gene_{x}\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1ed23423-489e-466c-9a1a-81787258b8a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shigella: 47\n"
     ]
    }
   ],
   "source": [
    "print(f\"Shigella: {len(genes_shi)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4f332e7-530f-4d3b-8313-f48401946abf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
